x-pack/filebeat/input/awscloudwatch/input.go

```diff
@@ -0,0 +1,360 @@
+				}
+			}
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	in.logger.Info("Initialized AWS CloudWatch input.")
+	return in, nil
+}
+
+// Run runs the input
+func (in *awsCloudWatchInput) Run() {
+	in.workerOnce.Do(func() {
+		in.workerWg.Add(1)
+		go func() {
+			in.logger.Infof("awscloudwatch input worker for log group: '%v' has started", in.config.LogGroup)
+			defer in.logger.Infof("awscloudwatch input worker for log group '%v' has stopped.", in.config.LogGroup)
+			defer in.workerWg.Done()
+			in.run()
+		}()
+	})
+}
+
+func (in *awsCloudWatchInput) run() {
+	cwConfig := awscommon.EnrichAWSConfigWithEndpoint(in.config.AwsConfig.Endpoint, "cloudwatchlogs", in.config.RegionName, in.awsConfig)
+	svc := cloudwatchlogs.New(cwConfig)
+	for in.inputCtx.Err() == nil {
+		err := in.getLogEventsFromCloudWatch(svc)
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == awssdk.ErrCodeRequestCanceled {
+				continue
+			}
+			in.logger.Error("getLogEventsFromCloudWatch failed: ", err)
+			continue
+		}
+
+		in.logger.Debugf("sleeping for %v before checking new logs", in.config.ScanFrequency)
+		time.Sleep(in.config.ScanFrequency)
+		in.logger.Debug("done sleeping")
+	}
+}
+
+func parseARN(logGroupARN string) (string, string, error) {
+	arnParsed, err := arn.Parse(logGroupARN)
+	if err != nil {
+		return "", "", errors.Errorf("error Parse arn %s: %v", logGroupARN, err)
+	}
+
+	if strings.Contains(arnParsed.Resource, ":") {
+		resourceARNSplit := strings.Split(arnParsed.Resource, ":")
+		if len(resourceARNSplit) >= 2 && resourceARNSplit[0] == "log-group" {
+			return resourceARNSplit[1], arnParsed.Region, nil
+		}
+	}
+	return "", "", errors.Errorf("cannot get log group name from log group ARN: %s", logGroupARN)
+}
+
+func (in *awsCloudWatchInput) getLogEventsFromCloudWatch(svc cloudwatchlogsiface.ClientAPI) error {
+	errC := make(chan error, 1)
+	cwCtx := &cwContext{
+		refs: 1,
+		errC: errC,
+	}
+	defer cwCtx.done()
+
+	ctx, cancelFn := context.WithTimeout(in.inputCtx, in.config.APITimeout)
+	defer cancelFn()
+
+	i := 0
+	nextToken := ""
+	currentTime := time.Now()
+	startTime, endTime := getStartPosition(in.config.StartPosition, currentTime, in.prevEndTime, in.config.ScanFrequency)
+	in.logger.Debugf("start_position = %s, startTime = %v, endTime = %v", in.config.StartPosition, time.Unix(startTime/1000, 0), time.Unix(endTime/1000, 0))
+
+	// overwrite prevEndTime using new endTime
+	in.prevEndTime = endTime
+
+	for nextToken != "" || i == 0 {
+		// construct FilterLogEventsInput
+		filterLogEventsInput := in.constructFilterLogEventsInput(startTime, endTime, i, nextToken)
+
+		// make API request
+		req := svc.FilterLogEventsRequest(filterLogEventsInput)
+		resp, err := req.Send(ctx)
+		if err != nil {
+			in.logger.Error("failed FilterLogEventsRequest", err)
+			return err
+		}
+
+		// get token for next API call
+		nextToken = ""
+		if resp.NextToken != nil {
+			nextToken = *resp.NextToken
+		}
```

Comment:
This should finish when the new nextToken is equal to the current one

Reply:
`FilterLogEvents` returns `""` when pagination is done.